feat: Implement the publishBatch hook for smart transactions

dan437 opened the pull request:
## **Description**
This PR implements the publishBatch hook for smart transactions, which will be used when submitting multiple transactions at once, e.g. when a user choose their gas token for a transaction.

We can test this end to end once UI for the Transaction Confirmation page is merged. Then we might need to do some final tweaks if needed.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/31267?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Once the UI on the Transaction Confirmation page is ready (e.g. when you do a Send transaction), you will be able to choose a gas token to be used for transaction fees

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

dan437 committed:
Implement the publishBatch hook for smart transactions

Use the PublishBatchHookRequest type

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

Make transactionMeta mandatory for submitBatch, export getTransactionById, refactoring

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 commented on app/scripts/lib/transaction/smart-transactions.ts:
```diff
@@ -58,14 +58,22 @@ export type FeatureFlags = {
   };
 };
 
+export type SignedTransaction = {
```
We could instead use a type from the TransactionController once its available.

dan437 commented on app/scripts/lib/transaction/smart-transactions.ts:
```diff
@@ -112,21 +122,26 @@ class SmartTransactionHook {
-    this.#txParams = transactionMeta.txParams;
-    this.#shouldShowStatusPage =
-      transactionMeta.type !== TransactionType.bridge;
+    this.#isDapp = transactionMeta?.origin !== ORIGIN_METAMASK;
```
TODO: Since we don't pass `transactionMeta` in the submitBatch hook, we need to update this. We can perhaps pass the `origin` or `isDapp` in the constructor.

dan437 replied:
Resolved by always passing `transactionMeta` into the hook.

dan437 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -1,5 +1,6 @@
 import {
   CHAIN_IDS,
+  type PublishBatchHookRequest,
```
This one will be available once the TransactionController changes are merged and used in the extension.

dan437 committed:
End an existing approval flow before starting a new one

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 committed:
Add a new test

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

dan437 committed:
Use type PublishBatchHookTransaction

Signed-off-by: dan437 <80175477+dan437@users.noreply.github.com>

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -153,6 +157,15 @@ export const TransactionControllerInit: ControllerInitFunction<
           rawTx,
         ),
+      publishBatch: async (_request: PublishBatchHookRequest) =>
+        await publishBatchSmartTransactionHook({
```
As these `-init` files are to decouple the controllers and simplify code ownership, it would be great in future if we could move "everything" into `lib/smart-transactions` and perhaps just invoke it here via a messenger call such as `SmartTransactions:publishHook`, or at least construct the class directly without all the logic below.

dan437 replied:
Nice idea, will create a ticket for moving STX specific code into lib/smart-transactions.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -153,6 +157,15 @@ export const TransactionControllerInit: ControllerInitFunction<
+      publishBatch: async (_request: PublishBatchHookRequest) =>
+        await publishBatchSmartTransactionHook({
+          transactionController: controller,
+          smartTransactionsController: smartTransactionsController(),
```
Same here, not needed now but ideally we'd avoid passing controller instances around, and rely on the messenger to communicate with them.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -247,6 +260,62 @@ function publishSmartTransactionHook(
+  const featureFlags = getFeatureFlagsByChainId(uiState);
+
+  // Get transactionMeta based on the last transaction ID
+  const lastTransaction = transactions[transactions.length - 1];
```
Could this logic be encapsulated inside the other file and we just pass in the ID?

dan437 replied:
The hook constructor expects `transactionMeta`, so I want to get it here before I call `submitBatchSmartTransactionHook`. I also want to throw an error early if there is no `transactionMeta`, so I think keeping it here makes sense.

matthewwalsh0 commented on app/scripts/controller-init/confirmations/transaction-controller-init.ts:
```diff
@@ -247,6 +260,62 @@ function publishSmartTransactionHook(
+}) {
+  // UI state is required to support shared selectors to avoid duplicate logic in frontend and backend.
+  // Ideally all backend logic would instead rely on messenger event / state subscriptions.
+  const uiState = getUIState(flatState);
```
This is very similar to the other hook, so pending a future refactor to just use the messenger, we could have a general function here to remove some duplication?

matthewwalsh0 commented on app/scripts/controller-init/messengers/transaction-controller-messenger.ts:
```diff
@@ -102,6 +102,7 @@ export function getTransactionControllerInitMessenger(
       'ApprovalController:endFlow',
       'ApprovalController:startFlow',
       'ApprovalController:updateRequestState',
+      'ApprovalController:acceptRequest',
```
Minor, alphabetical order.

matthewwalsh0 commented on app/scripts/lib/transaction/smart-transactions.ts:
```diff
@@ -210,14 +216,129 @@ class SmartTransactionHook {
+    const useRegularTransactionSubmit = undefined;
+
+    if (!this.#isSmartTransaction) {
+      return useRegularTransactionSubmit;
```
We don't yet have a fallback here, so could throw to have a more meaningful error than `No publish hook found` which is what we'd get currently.

Though naturally this flow shouldn't be possible at the moment.

matthewwalsh0 commented on app/scripts/lib/transaction/smart-transactions.ts:
```diff
@@ -210,14 +216,129 @@ class SmartTransactionHook {
+      }
+
+      let submitBatchResponse;
+      if (submitTransactionResponse?.txHashes) {
```
Minor, ternary instead of `let`?

dan437 replied:
I've tried that after your suggestion, but it seemed little less readable. Also, AI mentioned:
_While the ternary operator is more concise, when dealing with complex nested operations like the map function inside object literals, the traditional if/else structure makes the conditional logic clearer and easier to scan._

matthewwalsh0 commented on app/scripts/lib/transaction/smart-transactions.ts:
```diff
@@ -210,14 +216,129 @@ class SmartTransactionHook {
-      id: this.#approvalFlowId,
-    });
+    try {
+      this.#controllerMessenger.call('ApprovalController:endFlow', {
```
Minor, could have a function to encapsulate this as we do it multiple times.

dan437 commented:
Thanks a lot for great feedback, @matthewwalsh0 . It seems the PR got approved and automatically merged for today's release candidate, since it's very high priority. I will send a follow up PR which addresses your comments and get that cherry picked into the RC.
